Modify upstream RAID deletion to put all disks in JBOD mode after deleting raid configuration

Rachit7194 opened the pull request:
Hey @cdearborn  .
This changes includes realtime raid creation/deletion changes and reset_configuration implementation.
As this PR need to test. i have put this under WIP.
Also i need to discuss few issues i am facing to get clear.

Rachit7194 commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -924,13 +918,13 @@ def delete_configuration(self, task):
             return None
-        return node_state
 
+        return node_state
```
The reason i am returning None in above line i  am explaining below:
1. Convert all disk to JBOD(non-raid) mode, after deleting raid(VD, hotspares/dedicated hotspares).
2. Now currently `delete_configuration` function returns `clean_wait` after deleting virtual disks , its resumes cleaning and assign_roles converts all disks to jbod.
-------------Here comes new implementation-------------------
3. Now modifying `delete_configuration` call by applying  `reset_configuration` , it returns the same `clean_wait` but unfortunately it keeps giving node state in `clean_wait` instead of  resuming cleaning and then `manageable`.
3. So after reset_configuration call i applied call for conversion of disks to raid 
and when i return None from `delete_configuration()`,  it successfully converts all physical disks to `NON-RAID` 
4. I am not sure whether to return `None` or `clean_wait`
NOTE:  `clean_wait` returns by `_commit_to_controllers()`
Please let me know your review.

cdearborn replied:
As I understand it, you return CLEANWAIT if an operation executes asynchronously.  Meaning that the operation continues execution after the method returns.  You only return None if the operation has completed.
I'll look into this more on Monday.

Rachit7194 replied:
Yes actually, reset_configuration() and disk_conversion() call cannot call asynchronously.
because reset_configuration should be completed first and after that only i can check all disks which are not in NON-RAID and can convert into NON-RAID/JBOD mode

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -761,13 +910,27 @@ def delete_configuration(self, task):
+        controllers = list()
+        drac_raid_controllers = list_raid_controllers(node)
+        for cntrl in drac_raid_controllers:
+            if is_raid_controller(node, cntrl.id):
```
I think we have a bug in the existing code.  is_raid_controller will return False for BOSS cards, which is incorrect.  The end result is that the RAID1 on BOSS cards will not be deleted by this code.  Has the code below been tested on BOSS cards?

Rachit7194 replied:
No as of now we are testing only on Raid Controllers (H730 perc) which is present on 13G (R7)stamp

cdearborn replied:
This patch resolves the issue: https://review.opendev.org/#/c/653850/

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -202,20 +271,77 @@ def delete_virtual_disk(node, virtual_disk):
+
+    :param node: an ironic node object.
+    :param raid_controller: id of the RAID controller.
+    :returns: a dictionary containing the commit_needed key with a boolean
```
In the line above, I believe that key that will be returned is: "is_commit_required"

Rachit7194 replied:
Ack

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -202,20 +271,77 @@ def delete_virtual_disk(node, virtual_disk):
+
+    :param node: an ironic node object.
+    :param raid_controller: id of the RAID controller.
+    :returns: a dictionary containing the commit_needed key with a boolean
```
In the line above, I believe that key that will be returned is: "is_commit_required"

Rachit7194 replied:
Yes i will change and update in new commit..i have just injected reset_config() method to test raid/jbod disks conversion..

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -761,13 +910,27 @@ def delete_configuration(self, task):
+        node_state = _commit_to_controllers(node, controllers)
+        LOG.info("Converting all physical disks to JBOD mode")
+        change_state_result = configure_physical_disk_conversion(node,
+            mode="JBOD")
```
On the line above, I think we should pass in dracclient.constants.RaidStatus.jbod

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -202,20 +253,77 @@ def delete_virtual_disk(node, virtual_disk):
+
+
+def clear_foreign_config(node, raid_controller):
+    """Free up the foreign drivers.
```
Can you change "drivers" to "drives" in the line above?

Rachit7194 replied:
already changed..It will reflect into new commit

Rachit7194 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -134,6 +135,56 @@ def list_physical_disks(node):
+        change_state_result = client.change_physical_disk_state(
+            mode, controllers_to_physical_disk_ids)
+        controllers = []
+        if change_state_result['commit_required_ids']:
```
This code will need to be updated to use: https://review.opendev.org/#/c/654608/

Rachit7194 replied:
This function will not be needed. as you suggested new implementation to add substep after delete_configuration(). so i will remove this in next commit.

Rachit7194 commented:
Hi @cdearborn ,
Just want to update few cases for latest commit which includes substep implementation.
1. There is one typo mistake in parameter values of _commit_to_controllers(), i will update that in an hour.
2. Now there are three substep occured:
    i. delete_reset_configuration.
    ii. delete_foreign_configuration.
    iii. change_phy_disk_state
2. This substep execution does all the substep successfully but sometimes case happen that 
   when it checks for the last substep after completing (let say. `change_physical_disk_state`), we need to call 
   resume_cleaning(), but node goes into cleaning state infinitely and it throws below log many times: 

> `2019-05-06 13:35:49.153 4702 DEBUG ironic.drivers.modules.agent_base_vendor [-]
>  Heartbeat from node bc523467-e907-455e-86f1-ca79cd87238a in unsupported provision state cleaning,  not taking any action. heartbeat /usr/lib/python2.7/sitepackages/ironic/drivers/modules/agent_base_vendor.py:290
> `
Also calling _resuming_cleaning() method gives error on line https://github.com/cdearborn/ironic/blob/bf89ad0778007708fe902c08e9e3b61453eb9390/ironic/drivers/modules/drac/raid.py#L1021
Below is the error block:
```
`2019-05-06 13:26:21.908 4702 INFO ironic.drivers.modules.drac.raid [req-9d0da9f7-49fb-43bd-aaa0-52bb4de30094 - - - - -] Change has been committed to RAID controller {'is_reboot_required': True, 'raid_controller': u'RAID.Integrated.1-1'} on node bc523467-e907-455e-86f1-ca79cd87238a. DRAC job id: JID_571671787579
 ERROR futurist.periodics [req-9d0da9f7-49fb-43bd-aaa0-52bb4de30094 - - - - -] Failed to call periodic 'ironic.drivers.modules.drac.raid.DracRAID._query_raid_config_job_status' (it runs every 120.00 seconds): AttributeError: 'NoneType' object has no attribute 'driver_info'
 ERROR futurist.periodics Traceback (most recent call last):
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/futurist/periodics.py", line 290, in run
 ERROR futurist.periodics     work()
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/futurist/periodics.py", line 64, in __call__
 ERROR futurist.periodics     return self.callback(*self.args, **self.kwargs)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic_lib/metrics.py", line 60, in wrapped
 ERROR futurist.periodics     result = f(*args, **kwargs)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/futurist/periodics.py", line 178, in decorator
 ERROR futurist.periodics     return f(*args, **kwargs)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/raid.py", line 927, in _query_raid_config_job_status
 ERROR futurist.periodics     self._check_node_raid_jobs(task)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic_lib/metrics.py", line 60, in wrapped
 ERROR futurist.periodics     result = f(*args, **kwargs)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/raid.py", line 946, in _check_node_raid_jobs
 ERROR futurist.periodics     LOG.info("#"*90)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/raid.py", line 1074, in _resume_cleaning
 ERROR futurist.periodics     task.node, self.get_logical_disks(task))
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic_lib/metrics.py", line 60, in wrapped
 ERROR futurist.periodics     result = f(*args, **kwargs)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/raid.py", line 889, in get_logical_disks
 ERROR futurist.periodics     for disk in list_virtual_disks(node):
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/raid.py", line 109, in list_virtual_disks
 ERROR futurist.periodics     client = drac_common.get_drac_client(node)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/common.py", line 128, in get_drac_client
 ERROR futurist.periodics     driver_info = parse_driver_info(node)
 ERROR futurist.periodics   File "/usr/lib/python2.7/site-packages/ironic/drivers/modules/drac/common.py", line 63, in parse_driver_info
 ERROR futurist.periodics     driver_info = node.driver_info
 ERROR futurist.periodics AttributeError: 'NoneType' object has no attribute 'driver_info'
 ERROR futurist.periodics
```

3. Sometimes its easily pass all the condition return managable state where it completes delete_configurations and return to assign_role.py
4. Currently i have not tested this implementation on latest master branch of upstream ironic as i am stiill setting up keystone service on my VM

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -639,18 +716,25 @@ def _commit_to_controllers(node, controllers):
     for controller in controllers:
-        # Do a reboot only for the last controller
+        raid_controller = controller['raid_controller']
+        realtime = controller['is_reboot_required'] == 'optional'
```
On the above line, we should change 'optional' to RebootRequired.optional

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -659,6 +743,12 @@ def _commit_to_controllers(node, controllers):
+        if raid_controller not in driver_internal_info[
+            'raid_config_parameters']:
+            driver_internal_info['raid_config_parameters'].append(
+                raid_controller)
```
It looks like the if statement on line 747 above will always fail because the raid controller will already be in the list due to line 746.  As a result, it looks like lines 747-750 should be removed.

Rachit7194 replied:
Removed line 746. as discussed

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -134,6 +134,26 @@ def list_physical_disks(node):
         raise exception.DracOperationError(error=exc)
 
 
+def is_raid_controller(node, raid_controller_fqdd):
```
This will need to be updated for this patch: https://review.opendev.org/#/c/653850/

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -782,15 +867,18 @@ def delete_configuration(self, task):
-        return _commit_to_controllers(node, controllers)
+        drac_raid_controllers = list_raid_controllers(node)
+        for cntrl in drac_raid_controllers:
+            if is_raid_controller(node, cntrl.id):
```
You'll need to update the above call to pass the list of RAID controllers in so that we don't beat up the iDRAC with calls to enumerate DCIM_ControllerView.

Rachit7194 replied:
on line 870: the function is just a anchor or a bridge or simple wrapper which call the list_raid_controllers() from dracclient.

cdearborn replied:
Hey Rachit, I recently changed is_raid_controller() so that it will return True for both BOSS cards and PERC RAID Controllers (https://review.opendev.org/#/c/653850/).  As part of this change, I added an optional parameter to is_raid_controller() that allows you to pass in the list of controllers.  I did this because the code that checks to see if a controller is a BOSS controller has to issue a query to the iDRAC to get the list of controllers.  Since we frequently call is_raid_controller() in a loop as above, I didn't want us to be needlessly querying the iDRAC with repeated calls to get the list of controllers.  The end result is that you can pass in the list of controllers to is_raid_controller() and that will optimize things so that it will use the passed in list instead of querying the iDRAC.  To take advantage of this optimization, you will need to change line 872 above so that is passes the list of controllers to is_raid_controller(), and change that local method so that it passes the list along to client.is_raid_controller().  See https://github.com/openstack/python-dracclient/blob/master/dracclient/client.py#L1008

Rachit7194 replied:
Oh.! okay i got your point...Thanks

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
                 self._clear_raid_config_job_failure(node)
                 self._set_clean_failed(task, config_job)
 
+        if finished_job_ids and not node.driver_internal_info.get(
```
I think we should remove the "if" statement above and move the code inside this if statement to line 970 above.

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
                 self._clear_raid_config_job_failure(node)
```
This block of code, starting with "if not node.driver_internal_info['raid_config_job_ids']:" needs to be combined with the the if statement below on line 975.  As it is, when the config job finishes, the above code will call _resume_cleaning(), which is not what we want if there is another substep to execute.
I think we should leave the code above as-is, and move the contents of the "if" statement below to line 970 above.
I think that this is what's causing the issues that you're having.

Rachit7194 replied:
done.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
 
+        if finished_job_ids and not node.driver_internal_info.get(
+                'raid_config_job_failure'):
+            if 'raid_config_substep' in node.driver_internal_info:
```
There needs to be an else: clause for the if statement above that calls: self._resume_cleaning(task)
This is because we have no substep, and all the jobs are complete, so this cleaning step is done.

Rachit7194 replied:
This means we need to remove 'raid_config_substep' from node.driver_internal_info when it completes the last substep(change phy disk to JBOD).
Because if we add else cause in above if statement(line-977), 'raid_config_substep' should be removed, after than only it will execute self.resume_cleaning(task)

Rachit7194 replied:
Updated in new commit.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
+                if node.driver_internal_info['raid_config_substep'] == \
+                        'delete_reset_config':
+                    controllers = list()
+                    no_jobs_required = False
```
I think it's confusing to have a variable name begin with "no".  Can you change the above variable name to "jobs_required" and set it to False?

Rachit7194 replied:
done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
+                                          'is_reboot_required']}
+                        controllers.append(controller)
+                        no_jobs_required = no_jobs_required == controller_cap[
+                            'is_commit_required']
```
I think the above line should be:
jobs_required = jobs_required or controller_cap['is_commit_required']

Rachit7194 replied:
done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -884,6 +972,65 @@ def _check_node_raid_jobs(self, task):
+                        no_jobs_required = no_jobs_required == controller_cap[
+                            'is_commit_required']
+
+                    if no_jobs_required:
```
With the above changes, the line above will be:
if not jobs_required:

Rachit7194 replied:
done.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -134,6 +134,27 @@ def list_physical_disks(node):
         raise exception.DracOperationError(error=exc)
 
 
+def is_raid_controller(node, raid_controller_fqdd, raid_controllers=None):
```
Can you add an entry for the raid_controllers param in the comments below?

Rachit7194 replied:
Added

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -879,7 +970,64 @@ def _check_node_raid_jobs(self, task):
+                                                  'is_reboot_required']}
+                            controllers.append(controller)
+                            jobs_required = jobs_required == controller_cap[
+                                'is_commit_required']
```
I think the logic in the 2 lines above is still not right.  On the first pass through the loop, if is_commit_required is False, then jobs_required will be True, with is not correct.
I think what we need here is:
jobs_required = jobs_required or controller_cap['is_commit_required']
That way, if there is at least one controller that requires a commit, jobs_required will end up being True.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -879,7 +970,64 @@ def _check_node_raid_jobs(self, task):
+                            jobs_required = jobs_required == controller_cap[
+                                'is_commit_required']
+
+                        if jobs_required:
```
I think the above line needs to be:
if not jobs_required:
This will make it so that if there were no foreign drives, it will enter the code below, log the "No foreign drives" message and continue.

Rachit7194 replied:
Resolved this error

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -644,8 +719,15 @@ def _commit_to_controllers(node, controllers):
+    if 'raid_config_parameters' not in driver_internal_info:
+        driver_internal_info['raid_config_parameters'] = []
+    if driver_internal_info['raid_config_substep'] == 'complete':
+        del driver_internal_info['raid_config_substep']
```
Can you add a command here to delete the raid_config_parameters too?  I'm concerned that if we don't delete it, it will hold data that will affect the execution when the cleaning step is run again.

Rachit7194 replied:
Added  at the end of the function call because it was giving KeyError below while adding job_id

Rachit7194 commented:
Hi Chris,
I have added one additional substep that is 'virtual_disk_created' because at the time of virtual disk creation it was taking substep default value to None at line 864  `_commit_to_controllers(node, controllers)`, as it was not having substep parameter, Therefore it was not resuming to clean . 
So after adding substep='virtual_disk_created' it will call self._resume_cleaning() in check_node_raid_jobs
Also above additional substep i have not tested yet..i will test it by the end of the day

Also i want to know that can we create a single patch to upstream, because current PR looks like patch cannot be divided into reset_config, and substep flow implementation(which includes clear_foreing_confg, disk_conversion)..what do you think?

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -470,7 +549,7 @@ def _find_configuration(logical_disks, physical_disks):
                                         free_space_mb))
             if not result:
-                error_msg = _('failed to find matching physical disks for all '
+                error_msg = _('failed to find matching physical disks for all'
```
I think the space above needs to be added back in, otherwise there will be no space between "all" and "logical"

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -675,6 +760,15 @@ def _commit_to_controllers(node, controllers):
+
+    if driver_internal_info['raid_config_substep'] == 'completed':
+        del driver_internal_info['raid_config_substep']
+        del driver_internal_info['raid_config_parameters']
```
I don't think we can delete raid_config_substep and raid_config_parameters here.  That should be done from the periodic task.  Let's create a helper method that checks if these keys exist in driver_internal_info and deletes them if they do.  You could call it something like _clear_raid_substep().  Then call this method from the _check_node_raid_jobs method wherever resume_cleaning is called or a failure occurs.

Rachit7194 replied:
Yes but we cannot call `_clear_raid_substep()` in else cause of if condition(`if 'raid_config_substep' in node.driver_internal_info:`) at line.no- 973 where it checks key 'raid_config_substep' otherwise it will give KeyError. that ` 'raid_config_substep' not found in _clear_raid_substep()`.

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -644,8 +724,12 @@ def _commit_to_controllers(node, controllers):
+
     if 'raid_config_job_ids' not in driver_internal_info:
         driver_internal_info['raid_config_job_ids'] = []
+    if 'raid_config_parameters' not in driver_internal_info:
```
I think we should unconditionally set the raid_config_parameters to [].  This will make it so that we are guaranteed that there is no old data in it.

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -767,7 +861,8 @@ def create_configuration(self, task,
 
-        return _commit_to_controllers(node, controllers)
+        return _commit_to_controllers(node, controllers,
+                                      substep="virtual_disk_created")
```
Instead of adding a new substep, let's make the default for substep be "completed" and not specify it here.

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -635,7 +715,7 @@ def _filter_logical_disks(logical_disks, include_root_volume,
 
 
-def _commit_to_controllers(node, controllers):
+def _commit_to_controllers(node, controllers, substep=None):
```
Instead of defaulting substep to None, let's default it to "completed"

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +974,73 @@ def _check_node_raid_jobs(self, task):
+                        controllers, substep='completed')
+
+                if node.driver_internal_info['raid_config_substep'] == \
+                        'completed' or "virtual_disk_created":
```
If we make the default for substep be "completed" in the _commit method, then you will be able to remove "or virtual_disk_created" on the line above.

Rachit7194 replied:
Done

cdearborn commented:
Yes, I think this can be submitted as one patch upstream, though we'll need to make sure we create a good commit message that details how it frees foreign config drives and spares as well as converts drives from RAID to JBOD on delete and JBOD to RAID on create.

cdearborn commented:
Are you planning on making another PR for moving the required drives into RAID mode automatically during RAID volume creation?

Rachit7194 commented:
I can create another PR  or create commit in same PR for moving drives back to raid before raid creation.
As u suggest

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -470,8 +549,8 @@ def _find_configuration(logical_disks, physical_disks):
             if not result:
-                error_msg = _('failed to find matching physical disks for all '
-                              'logical disks')
+                error_msg = _('failed to find matching physical disks'
```
Need to add a space after "disks" on the line above, or this will end up being "disksfor"

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +968,84 @@ def _check_node_raid_jobs(self, task):
+                        controllers, substep='completed')
+
+                if node.driver_internal_info['raid_config_substep'] == \
+                        'completed':
```
I think what's happening is that the "change_disk_state" substep is executing, and then execution continues here.  This "if" statement succeeds because the change_disk_state substep just set the  raid_config_substep to completed, and then the 2 lines below are executed.  All of this happens before the conversion job has a chance to complete.
To fix this, I would changed the "if" statement above to be an "elif" statement.  This will prevent this code from being immediately executed, and instead it will be executed after the job completes.

Rachit7194 replied:
Done the change.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +968,84 @@ def _check_node_raid_jobs(self, task):
+                            substep='change_disk_state')
+
+                if node.driver_internal_info['raid_config_substep'] == \
+                        'change_disk_state':
```
I would change the "if" statement above to be an "elif".

Rachit7194 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +968,84 @@ def _check_node_raid_jobs(self, task):
+                        'change_disk_state':
+                    # In change_phy_disks state to jbod after clear reset
+                    # configuration and deleting foreign drives if any
+
```
I would factor out the code inside this "if" statement into a separate method and call it from here.

Rachit7194 replied:
Correct. done

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +968,84 @@ def _check_node_raid_jobs(self, task):
+                            "No foreign drives detected,"
+                            "performing next substep : {}".format(
+                                node.driver_internal_info[
+                                    'raid_config_substep']))
```
Instead of relying on the falling through to the next "if" statement and converting the drives, I would call into the factored out method directly here.  (See below comment)

Rachit7194 replied:
Nice idea. done

Rachit7194 commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -876,13 +968,84 @@ def _check_node_raid_jobs(self, task):
+                if node.driver_internal_info['raid_config_substep'] == \
+                        'completed':
+                    self._clear_raid_substep(node)
+                    self._resume_cleaning(task)
```
Yes i just tried removing self.resume_cleaning() on this line and it  executed successfully but i  will changed the flow as you changes which seems more  accurate.

Rachit7194 commented:
I have test last change, it successfully executed the substep from delete-configuration to create_configuration on single node but i need to test this implementation on 2-3 nodes more.

Rachit7194 commented:
I just tested the changes on another two nodes and the result what i saw is that :
- During `delete_configuration`  it executes :
     1. reset_raid_config ()
     2. substep_flow(checks foreign_drives, phy_disk_conversion to jbod)
     3. clean_step resume successfully
- Then assign_role converts physical disks back to raid 
- Set target_raid_config on node
- After create_configuration() clean_step get called
- Here i notice that when it calls _commit_to_controller() to apply changes at line [#885](https://github.com/cdearborn/ironic/blob/a474958d2c4a730f04b774d08d3f6fa5c64136d7/ironic/drivers/modules/drac/raid.py#L885) node directly returns from `clean_wait` to `cleaning` without creating a job 
I tried to search error related to this issue but i was not able find any, after that check `last_error` for node in database it was showing  this `Async execution of _do_next_clean_step failed with error: `.

Here i assume few failing issues.
logical_disks_create is empty when it iterates and returns None at line [#L869](https://github.com/cdearborn/ironic/blob/a474958d2c4a730f04b774d08d3f6fa5c64136d7/ironic/drivers/modules/drac/raid.py#L869). 

Though i am testing with different workarounds, i will update you on this in a while.

Rachit7194 commented:
Hey Chris,
Just an update, I had applied (pdb) debugger and logs on several scripts(manager.py, raid.py, assign_role.py) to test the execution while testing yesterday,
Now today i removed all debuggers and logs from the scripts and tested again on all 3 compute nodes,
and now its executing succesfully.
I will start working on test cases and update PR.

Rachit7194 commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -842,6 +846,29 @@ def create_configuration(self, task,
+
+        # Converting physical disks to raid before Raid Creation
+        _change_physical_disk_conversion(
+            node, mode="RAID",
```
i will add here mode=drac_common.drac_constants.RaidStatus.raid

Rachit7194 commented:
In function _commit_to_controller() , if i asssign value for `realtime =  drac_common.drac_constants.RebootRequired.optional` ,
i receive below error in test_raid.py: 
 `AttributeError: Mock object has no attribute 'RebootRequired'`
i cannot figure out why mock object does not have `RebootRequired`

cdearborn commented:
Hey Rachit, I think what's happening is that dracclient ends up referring to a mock, and the mock doesn't contain the RebootRequired class.  Instead of referencing drac_common.drac_constants, I think I would try importing dracclient.constants directly in raid.py.  See power.py for the approach used there.  It doesn't look to me like drac_common.drac_constants is used anywhere, so I'm guessing that it was added to use in power.py, but the author ran into the same problem that you just did and forgot to remove it from drac_common.

Rachit7194 commented:
Thanks. I will update it accordingly.

Rachit7194 commented:
Hey Chris, I tried to import RebootRequired class in raid.py but still `Mock` is giving me same attribute error in test_raid.py.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -926,19 +1112,37 @@ def test_create_configuration_with_backing_physical_disks(
                        autospec=True)
     def test_create_configuration_with_predefined_number_of_phyisical_disks(
-            self, mock_commit_config, mock_validate_job_queue,
+            self,  mock_change_physical_disk_state,
```
The order of the parameters in the method signature needs to be in reverse order of the decorators above, so the order of the parameters in the method signature should be:
            self, mock_commit_config, mock_validate_job_queue,
            mock_wait_for_jobs_completion, mock_list_unfinished_jobs,
            mock_list_physical_disks, mock_get_drac_client

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -926,19 +1112,37 @@ def test_create_configuration_with_backing_physical_disks(
     @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
                        autospec=True)
     @mock.patch.object(drac_raid, 'list_physical_disks', autospec=True)
+    @mock.patch.object(drac_raid, 'change_physical_disk_state', spec_set=True,
```
change_physical_disk_state is not a method in drac_raid, which is the same as ironic.drivers.modules.drac.raid.  As a result, the line above and below should be removed, and the "mock_change_physical_disk_state" parameter should be removed from the method signature.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -948,6 +1152,13 @@ def test_create_configuration_with_predefined_number_of_phyisical_disks(
         physical_disks = self._generate_physical_disks()
         mock_list_physical_disks.return_value = physical_disks
         mock_commit_config.side_effect = ['42', '12']
+        mock_change_physical_disk_state.return_value = {
```
Due to the changes recommend above, the above line should be changed to:
mock_client.change_physical_disk_state.return_value = {

Rachit7194 commented:
Hi Chris, 
I have updated last commit with only `delete_configuration` which includes substep implementation.
I only error i was receiving is `RebootRequired` Attribute not present in `Mock` object, so currently i have removed with constant variable.
raid.py line : 807
test_raid.py line : 1390
Also this change is pass tox(py27,py3,pep8) test
I will attach commit message here in an hour so that after your call i can raise a patch in upstream

Rachit7194 commented:
**### Updated Commit Message For Upstream Patch**:

> This patch upgrades Delete_configuration by replacing delete_virtual_disk with reset_raid_config which is responsible for removing Virtual Disks as well as Global and Dedicated Hot-Spares too.
> Also added a new Substep implementation followed by Cleaning Step. The main aim for this substep implementation is to execute multiple raid configuration jobs by executing single cleaning step. It can be accomplished by setting new parameter in node.driver_internal_info to keep the track of the current execution substep.
> This substep initiates when periodic task executes _query_raid_config_job_status() which checks whether the substep parameter in node.driver_internal_info is set or not. If set then it will execute the job related to substep or else resume cleaning.

Please let me know if i need to change the message

Rachit7194 commented:
As discussed, Added patch which includes freeing up spares and clearing foreign drives through substep infrastructure only

Rachit7194 commented:
Hi Chris, I Updated latest commit with resolved error(error was `AttributeError: Mock object has no attribute 'RebootRequired'` )

Though i was having having doubt at line 1448 of test_raid.py where it mock_commit_config taking params are realtime and reboot.

If you can review def test_delete_configuration() call in test_raid.py, i think all others test_cases will get reviewed automatically

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -202,6 +228,59 @@ def delete_virtual_disk(node, virtual_disk):
+
+    :param node: an ironic node object.
+    :param raid_controller: id of the RAID controller.
+    :returns: a dictionary containing the is_commit_required needed key with a
```
The dictionary will contain both is_commit_required and is_reboot_required.  Can you update the comment to include both?

Rachit7194 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -202,6 +228,59 @@ def delete_virtual_disk(node, virtual_disk):
+
+    :param node: an ironic node object.
+    :param raid_controller: id of the RAID controller.
+    :returns: a dictionary containing the is_commit_required needed key with a
```
The dictionary will contain both is_commit_required and is_reboot_required. Can you update the comment to include both?

Rachit7194 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -606,14 +606,17 @@ def _generate_physical_disks(self):
     @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
                        autospec=True)
+    @mock.patch.object(drac_common, 'get_reboot_options', spec_set=True,
+                       autospec=True)    
```
I think there's some trailing whitespace on line 610 that needs to be removed.

Rachit7194 replied:
Done.

rpioso commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1380,48 +1406,47 @@ def test_create_configuration_fails_if_disk_already_reserved(
-                task.node, raid_controller='RAID.Integrated.1-1', reboot=False,
-                realtime=True)
+                task.node, raid_controller='RAID.Integrated.1-1', reboot=True,
+                realtime=False)
```
Please remember to revert this change after addressing the issue with `mock_get_reboot_options`.

Rachit7194 replied:
Updated with latest commit.

rpioso commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1380,48 +1406,47 @@ def test_create_configuration_fails_if_disk_already_reserved(
+                                  mock_get_reboot_options,
                                   mock_get_drac_client):
         mock_client = mock.Mock()
         mock_get_drac_client.return_value = mock_client
```
I believe `mock_get_reboot_options`'s return value needs to be set to an object that looks like `dracclient.constants.RebootRequired`, analogous to how `mock_get_drac_client`'s is here. The following worked for me.

    mock_get_reboot_options.return_value = DictToObj(
        {'true': 'true', 'optional': 'optional', 'false': 'false'})

Rachit7194 replied:
Thanks Richard for resolving the issue. it passed the testcases.
Updated the change with latest commit.

cdearborn commented:
Hey Rachit, Rick and I spent quite a bit of time today talking about this, investigating it and coming up with a solution.  I'll send you an email with all we found out, along with a proposed solution.

Rachit7194 commented:
> Hey Rachit, Rick and I spent quite a bit of time today talking about this, investigating it and coming up with a solution. I'll send you an email with all we found out, along with a proposed solution.

Thank you Chris and Richard for helping out to mock RebootRequired properly

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1430,31 +1426,30 @@ def test_delete_configuration(self, mock_delete_virtual_disk,
-        mock_delete_virtual_disk.return_value = {
+        mock_list_raid_controllers.return_value = []
+        mock_reset_raid_config.return_value = {
             'is_reboot_required': 'optional',
```
I think 'optional' on the line above needs to be replaced with constants.RebootRequired().optional

cdearborn commented:
Closing this one since these changes have been submitted upstream:
https://review.opendev.org/#/c/664633/
